### PR TITLE
Containerize application and configure GitHub Actions to build and push Docker image to Harbor

### PR DESCRIPTION
**Pull request by VKTB**
## Description
This PR adds a Dockerfile which can be used to build a Docker image and create a container using it. It is configured to create a production image and run an Apache httpd server on port `80` when a container is started.

To add an extra layer of security, the image has been set up so that containers run as non-root.

Environment variables have also been defined in the Dockerfile to allow for values to be passed at runtime to future running containers. These values are then used by the `docker-entrypoint.sh` script, which acts as an entry point script that is executed when the container starts, to update the config values in the `datagateway-dataview-settings.json`, `datagateway-download-settings.json`, and `datagateway-search-settings.json` files. The environment variables are:
- `API_URL` (Default value: `/datagateway-api`)
- `DOWNLOAD_API_URL` (Default value: `http://localhost`)
- `ICAT_URL` (Default value: `http://localhost`)
- `IDS_URL` (Default value: `http://localhost`)

Finally, this PR adds a `docker` job to the `ci-build` GitHub Actions workflow to build a Docker image using the Dockerfile introduced in this PR and push the same to Harbor. The job only triggers if all the other jobs succeed and does different things depending on the context. It builds the Docker image in all cases and also pushes the image to Harbor only if something is pushed to the `main` or `develop` branch.

To build an image, run the command below in the `datagateway` repo directory
```
docker build -t datagateway_image .
```
To start a container on port `80` from the image that you just built, run:
```
docker run -p 80:80 --name datagateway_container datagateway_image 
```
If you want to pass values for the environment variables then instead run:
```
docker run -p 80:80 --name datagateway_container --env API_URL=/api --env ICAT_URL=https://127.0.0.1:8181 datagateway_image
```

## Testing instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] GitHub Actions should build and push an image to Harbor when this PR is merged

### CODE REVIEW

**Review by ajkyffin — DISMISSED** [review]

**Review by louise-davies — DISMISSED** [review]
LGTM - I deployed this on a cloud machine and it worked fine other than the issue I reported on Gitlab

**Comment by VKTB** [comment]
@louise-davies I fixed the caching issue as part of the last 2 commits.

**Review by louise-davies — COMMENTED** [review]
The caching doesn't work when loaded via the docker compose file (where everything is routed via a nginx server). Might only need modifications in that file which is in a separate repo, hence not necessicarily requesting changes on this PR

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+/packages/datagateway-*/cypress.json
+/packages/datagateway-*/server
+/packages/datagateway-*/README.md
+.gitignore
+.prettierrc
+codecov.yml
+CODEOWNERS
+LICENSE.md
+README.md
+Dockerfile
```
Comment (VKTB) [inline .dockerignore:13]:
```suggestion
/.github
/.husky
/packages/datagateway-*/cypress
/packages/datagateway-*/cypress.json
/packages/datagateway-*/server
/packages/datagateway-*/README.md
**/build
**/coverage
**/lib
**/node_modules
.gitignore
.prettierrc
codecov.yml
CODEOWNERS
LICENSE.md
README.md
Dockerfile
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,75 @@
+# Dockerfile to build and serve datagateway
+
+# Build stage
+FROM node:16.14-alpine3.15 as builder
```
Comment (louise-davies) [inline Dockerfile:4]: Node 16 is now no longer supported, Node 20 is current LTS and is what we're now running CI on.

I forgot to mention this for the SG PR, but can we test these containers with the up-to-date LTS version of Node?

Comment (VKTB) [reply Dockerfile:4]:
```suggestion
FROM node:20.9-alpine3.18 as builder
```

Comment (VKTB) [reply Dockerfile]: I updated it to Node 20 in https://github.com/ral-facilities/datagateway/pull/1535/commits/f9c87e00c8019a4deb8f900d069de38874cd6186 and it seems to be working fine. Feel free to test it.

**Review by louise-davies — APPROVED** [review]

**Pull request merged**